File: graph/Traversal/bfs.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

// BFS function
vector<int> bfsOfGraph(int V, vector<int> adj[]) {
    vector<int> bfs;
    vector<bool> visited(V, false);
    queue<int> q;

    visited[0] = true;  // Assuming BFS starts from node 0
    q.push(0);

    while (!q.empty()) {
        int node = q.front();
        q.pop();
        bfs.push_back(node);

        for (int neighbor : adj[node]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                q.push(neighbor);
            }
        }
    }
    return bfs;
}

int main() {
    int V, E;
    cout << "Enter number of vertices (V): ";
    cin >> V;

    cout << "Enter number of edges (E): ";
    cin >> E;

    vector<int> adj[V];

    cout << "Enter edges (u v): " << endl;
    for (int i = 0; i < E; ++i) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u); // Remove this line if the graph is directed
    }

    vector<int> result = bfsOfGraph(V, adj);

    cout << "BFS Traversal starting from node 0: ";
    for (int node : result) {
        cout << node << " ";
    }

    return 0;
}

File: graph/Traversal/bfs.py
from collections import deque

def bfs_of_graph(V, adj):
    visited = [False] * V
    bfs_result = []
    q = deque()

    visited[0] = True  # Starting from node 0
    q.append(0)

    while q:
        node = q.popleft()
        bfs_result.append(node)

        for neighbor in adj[node]:
            if not visited[neighbor]:
                visited[neighbor] = True
                q.append(neighbor)

    return bfs_result

# Taking user input
V = int(input("Enter number of vertices (V): "))
E = int(input("Enter number of edges (E): "))

adj = [[] for _ in range(V)]

print("Enter edges (u v):")
for _ in range(E):
    u, v = map(int, input().split())
    adj[u].append(v)
    adj[v].append(u)  # Remove this line if the graph is directed

# Call BFS
result = bfs_of_graph(V, adj)

# Output
print("BFS Traversal starting from node 0:")
print(*result)
